### ISSUE TITLE
Training models on SPICE

### ISSUE THREAD

**Issue by peastman** [openmm/spice-dataset#24]
We still have some time before data generation is complete, but I think this is a good time to start discussing the next stage of the project: creating ready-to-use models trained on SPICE. Here are some of the things we need to discuss.

What architectures do we want to try? My tests so far have used the equivariant transformer in torchmd-net.  What other architectures should we try out?

A related question is what methods of handling charges we want to try?  In my tests so far, I've done it by assigning a different atom type to each combination of element and formal charge.  There are other methods we could try as well, including ones that predict partial charges and include an explicit Coulomb term.

We need to create a test set. We could, of course, just pick a random set of samples from the original dataset to use as our test set, but I think we can do better. I suggest creating several new sets of conformations to separately test how it generalizes in different ways:

- New conformations for some molecules present in the training set.
- Molecules not present in the training set.
- Molecules that are larger than anything in the training set: larger drug molecules, longer peptides, water droplets.
- New classes of molecules that are not present in the training set: nucleic acids, lipids, carbohydrates, etc.

**Comment by peastman** [comment]
I've been running tests training models on a subset of the data.  Here are the architectures I've tried.

- Equivariant transformer
- NequIP
- Allegro

An issue with the latter two, which are both implemented in the same code, is that it requires the whole dataset to be loaded into memory at once.  That limits me to about 250,000 samples.  I've been talking to the developers about how to solve that problem.

All three models generally seem to be able to learn the data.  The best result I've gotten so far is with equivariant transformer, where I've gotten the RMS error on the validation set down to about 10 kJ/mol.  On the other two I've only gotten it down to about 30 kJ/mol.  Don't read anything into that though.  I've only tried default settings on them so far and haven't done any hyperparameter tuning at all.  The important message is that all three architectures basically work.

I also don't see a lot of sign of overfitting.  The validation set error is only slightly worse than the training set error.  That's encouraging, given that this only included three of the six PubChem sets.

To understand the results a bit better, I computed errors for 100,000 randomly chosen samples.  Here are the median, mean, and RMS errors (in kJ/mol)

median: 5.73
mean: 8.18
RMS: 12.85

If I exclude the 1% of samples with the largest errors, they get significantly better:

median: 5.65
mean: 7.54
RMS: 10.21

Given that, and given the large difference between median and RMS, it seems a small fraction of samples with large errors are bringing up the averages.  Molecule size is of course one factor affecting the error, but it doesn't seem to be the main one:

![image](https://user-images.githubusercontent.com/4379786/171061129-1b29f5e3-4435-4f83-aa0d-2ff8e20bf083.png)

In fact, if we divide the error by sqrt(# atoms) the result is basically flat, suggesting the per-atom errors are uncorrelated and independent of molecule size:

![image](https://user-images.githubusercontent.com/4379786/171061301-24b2ae4b-9a14-4928-add4-fb9cfb2ae3af.png)

Here are the SMILES strings for the ten samples with the largest errors.  All of these have errors >189 kJ/mol.  Note that several molecules appear multiple times.

```
[H:14][C:3]#[C:4][C:5]#[C:6][C:9](=[C:2]=[C:10]([H:16])[C:12]1([C:13]([C:11]([C:8](=[O:1])[O:7]1)([H:17])[H:18])([H:20])[H:21])[H:19])[H:15]
[H:14][C:3]#[C:4][C:5]#[C:6][C:9](=[C:2]=[C:10]([H:16])[C:12]1([C:13]([C:11]([C:8](=[O:1])[O:7]1)([H:17])[H:18])([H:20])[H:21])[H:19])[H:15]
[H:13][C:7]([H:14])([H:15])[S:4][C:8]([H:16])([H:17])[C:9]([H:18])([H:19])[C:5](=[N:1][H:10])[P:6]([O:2][H:11])[O:3][H:12]
[H:11][C:8]1=[N:4][C:9]2=[N:5][N:2]=[N:3][C:7](=[O:1])[C:10]2=[N:6]1
[H:10][C:9]([C:8](=[C:6]=[C:7]([Cl:1])[Cl:2])[Cl:3])([Cl:4])[Cl:5]
[H:11][C:3]1=[C:1]=[C:5]([C:9]([C:10]([C:6](=[C:2]=[C:4]([C:8]([C:7]1([H:15])[H:16])([H:17])[H:18])[H:12])[H:14])([H:21])[H:22])([H:19])[H:20])[H:13]
[H:11][C:8]1=[N:4][C:9]2=[N:5][N:2]=[N:3][C:7](=[O:1])[C:10]2=[N:6]1
[H:11][C:3]1=[C:1]=[C:5]([C:9]([C:10]([C:6](=[C:2]=[C:4]([C:8]([C:7]1([H:15])[H:16])([H:17])[H:18])[H:12])[H:14])([H:21])[H:22])([H:19])[H:20])[H:13]
[H:11][C:8]1=[N:4][C:9]2=[N:5][N:2]=[N:3][C:7](=[O:1])[C:10]2=[N:6]1
[H:11][C:3]1=[C:1]=[C:5]([C:9]([C:10]([C:6](=[C:2]=[C:4]([C:8]([C:7]1([H:15])[H:16])([H:17])[H:18])[H:12])[H:14])([H:21])[H:22])([H:19])[H:20])[H:13]
```

For comparison, here are the ten with the smallest errors.  All of these have errors <0.001 kJ/mol.

```
[H:17][N:15]([H:18])[C:7](=[O:1])[c:13]1[c:12]([c:14]([c:11]([c:9]([c:10]1[Cl:4])[Cl:3])[Cl:5])[C:8](=[O:2])[N:16]([H:19])[H:20])[Cl:6]
[H:20][C:12](=[C:11]([H:19])[C:13](=[C:15]1[C:9](=[N:3][O:4][C:7]1=[O:1])[C:17]([H:25])([H:26])[H:27])[H:21])[C:14]2=[C:10]([O:6][N:2]=[C:8]2[C:16]([H:22])([H:23])[H:24])[O:5][H:18]
[H:20][c:4]1[c:5]([c:7]([c:9]([c:8]([c:6]1[H:22])[H:24])[O:3][C:18]([C:13]([H:32])([H:33])[H:34])([C:14]([H:35])([H:36])[H:37])[C:17]([C:12]([H:29])([H:30])[H:31])([N:10]([H:25])[C:16]([H:40])([H:41])[C:15]([H:38])([H:39])[O:1][C:11]([H:26])([H:27])[H:28])[O:2][H:19])[H:23])[H:21]
[H:37][C:22]1([C:10](=[N:5][N:17]([C:8](=[O:2])[C:21]1([H:35])[H:36])[C:19]([H:30])([H:31])[C:9]2=[N:4][C:13]3=[C:12]([N:15]2[H:26])[C:11](=[C:14]([C:7](=[O:1])[N:16]3[C:18]([H:27])([H:28])[H:29])[C:23]([H:39])([H:40])[C:24]([H:41])([H:42])[C:20]([H:32])([H:33])[H:34])[O:6][H:25])[Cl:3])[H:38]
[H:11][C:6]1([C:4]([N:3]([C:5]1([H:9])[H:10])[C:2]#[N:1])([H:7])[H:8])[H:12]
[H:13][N:11]([H:14])[c:9]1[c:8]([c:10]([c:7]([c:5]([c:6]1[F:2])[F:1])[F:3])[N:12]([H:15])[H:16])[F:4]
[H:10][C:3]1([C:5]([C:9]2([C:6]([C:4]([C:8]1([C:7]2([H:18])[H:19])[H:20])([H:12])[H:13])([H:16])[H:17])[N+:2]#[C-:1])([H:14])[H:15])[H:11]
[H:14][C:5]1=[N:1][C:7]2=[C:8]([N:12]1[H:18])/[C:9](=[N:3]\\[O:4][C:13]([H:19])([H:20])[H:21])/[N:2]=[C:6]([N:11]2[H:17])[N:10]([H:15])[H:16]
[H:20][C:12](=[C:11]([H:19])[C:13](=[C:15]1[C:9](=[N:3][O:4][C:7]1=[O:1])[C:17]([H:25])([H:26])[H:27])[H:21])[C:14]2=[C:10]([O:6][N:2]=[C:8]2[C:16]([H:22])([H:23])[H:24])[O:5][H:18]
[H:17][C:6](=[N+:4]=[N-:1])[C:5](=[O:2])[C:12]1([C:15]2([C:8]([C:13]3([C:7]([C:14]([C:9]2([H:22])[H:23])([C:11]([C:16]1([C:10]3([H:24])[H:25])[Br:3])([H:26])[H:27])[H:30])([H:18])[H:19])[H:29])([H:20])[H:21])[H:31])[H:28]
```

One thing that does not seem to account for the difference is the presence of charges.  I was worried that my simple method of handling them (using different atom types for different formal charges) would lead to large errors on charged or polar molecules.  But none of the samples with the largest errors have any nonzero charges, whereas a couple of the samples with the lowest errors do.

**Comment by tmarkland** [comment]
Thanks Peter. For the molecules with the biggest errors, it might be worth checking the electronic structure output and/or recomputing the energies using a different code since although ground state DFT of these types of molecules doesn't fail that often it does sometimes happen and with a data set this large it wouldn't be surprising to me if some calculations had issues.

**Comment by peastman** [comment]
I looked a bit more into the molecules with very large errors.  Here's a plot comparing the reference and predicted energies for `[H:11][C:8]1=[N:4][C:9]2=[N:5][N:2]=[N:3][C:7](=[O:1])[C:10]2=[N:6]1`.  Conformations are sorted by error, so the most accurate ones are on the left and the least accurate ones on the right.

![image](https://user-images.githubusercontent.com/4379786/171213291-24e9f33c-8d53-4aa9-89e1-ece943587a74.png)

There are only a few conformations where it's wildly inaccurate.  For most conformations, the energies are consistently a little too large but overall not too bad.  With more data and/or training, it's likely the model could learn this.  Also notice that the very inaccurate conformations are the highest energy ones.  Maybe that's why the predictions on other conformations are consistently too high: it's compensating to try to get the highest energy ones not quite as far off?

Now compare to the energies for `[H:11][C:3]1=[C:1]=[C:5]([C:9]([C:10]([C:6](=[C:2]=[C:4]([C:8]([C:7]1([H:15])[H:16])([H:17])[H:18])[H:12])[H:14])([H:21])[H:22])([H:19])[H:20])[H:13]`.

![image](https://user-images.githubusercontent.com/4379786/171213828-6254d534-1890-4763-a948-1367932efaaf.png)

In this case, most energies are totally wrong.  There are about 15 conformations at the left where it's not so bad, but it isn't anywhere close to having learned to predict this molecule.  Also notice that there's no correlation between energy and error.  Unlike the previous molecule, very low energy conformations are just as likely to be predicted incorrectly as very high energy ones.

**Comment by giadefa** [comment]
Can you share the TorchMD-ET input file?

**Comment by peastman** [comment]
The config file is attached.

[config.yaml.txt](https://github.com/openmm/spice-dataset/files/8809357/config.yaml.txt)

**Comment by peastman** [comment]
I've been investigating this further with @YuezhiMao and @tmarkland.  A first possibility we needed to eliminate was that the DFT energies were incorrect, for example because it converged to an excited state rather than the ground state, or perhaps because of a bug in Psi4.  To test that, Yuezhi repeated the calculations for the second molecule above using Q-Chem with a few different methods: B3LYP-D3(BJ), wB97M-V, and RI-MP2.  The agreement was excellent.  So it doesn't seem to be a problem with the reference data.  The model just had trouble learning certain molecules.

I identified every sample for which the error was >100 kJ/mol.  There are 146 such samples out of more than 100,000 that I computed, so less that 0.15%.  For comparison, more than 10% of samples have errors below 1 kJ/mol.  Furthermore, those 146 samples are for just 30 molecules, and many of those have only 1 or 2 samples with errors above 100 kJ/mol.  There are 17 molecules with more than 2 conformations with very large errors.  Here they are.

```
[H:11][C:3]1=[C:1]=[C:5]([C:9]([C:10]([C:6](=[C:2]=[C:4]([C:8]([C:7]1([H:15])[H:16])([H:17])[H:18])[H:12])[H:14])([H:21])[H:22])([H:19])[H:20])[H:13]
[H:17][c:7]1[c:9]([c:13]2=[N:3][S:6][N:4]=[c:14]3[c:16]2[c:15]4[c:11]1=[N:1][S:5][N:2]=[c:12]4[c:8]([c:10]3[H:20])[H:18])[H:19]
[Na+:2].[I-:1]
[H:8][C:2](=[C:1]=[C:3]([C:5](=[C:4]([H:10])[H:11])[C:7]([H:15])([H:16])[H:17])[C:6]([H:12])([H:13])[H:14])[H:9]
[H:6][C:4](=[C:2]=[C:5]([H:8])[C:3]#[N:1])[H:7]
[H:14][C:3]#[C:4][C:5]#[C:6][C:9](=[C:2]=[C:10]([H:16])[C:12]1([C:13]([C:11]([C:8](=[O:1])[O:7]1)([H:17])[H:18])([H:20])[H:21])[H:19])[H:15]
[H:10][C:2](=[C:1]=[C:3]([C:8]1([C:4]([C:5]1([H:14])[H:15])([H:12])[H:13])[H:20])[C:9]2([C:6]([C:7]2([H:18])[H:19])([H:16])[H:17])[H:21])[H:11]
[H:4][C:3](=[S:2]=[O:1])[H:5]
[H:16][C:6](=[C:3]=[C:7]([H:17])[C:15]([C:10]([H:24])([H:25])[H:26])([C:11]([H:27])([H:28])[H:29])[C:14]([H:34])([H:35])[C:13]([H:32])([H:33])[C:12]([H:30])([H:31])[C:5](=[O:2])[C:9]([H:21])([H:22])[H:23])[C:4](=[O:1])[C:8]([H:18])([H:19])[H:20]
[H:8][C:5]([H:9])([H:10])[C:3](=[S:2]=[O:1])[N:4]([H:6])[H:7]
[H:16][c:6]1[c:7]([c:9]([c:11]2=[C:5]([C:3](=[C:1]=[C:2]([H:12])[H:13])[C:4](=[c:10]2[c:8]1[H:18])[H:14])[H:15])[H:19])[H:17]
[H:10][C:9]([C:8](=[C:6]=[C:7]([Cl:1])[Cl:2])[Cl:3])([Cl:4])[Cl:5]
[O:5]([P:6]([F:1])[F:2])[P:7]([F:3])[F:4]
[H:18][c:8]1[c:7]([c:9]([c:11]2[c:12]([c:10]1[H:20])[C:6](=[N:2][O:4]2)[C:13]([H:21])([H:22])[S:14](=[N:3][H:15])(=[O:1])[O:5][H:16])[H:19])[H:17]
[C:5](=[C:6]([F:1])[F:2])=[C:7]([F:3])[F:4]
[H:18][c:8]1[c:9]([c:11]([c:13]([c:12]([c:10]1[H:20])[H:22])[C:6](=[C:5]=[C:7]([H:17])[S:14](=[O:1])[C:15]([Cl:2])([Cl:3])[Cl:4])[H:16])[H:21])[H:19]
[H:11][C:8]1=[N:4][C:9]2=[N:5][N:2]=[N:3][C:7](=[O:1])[C:10]2=[N:6]1
```

The motif =C=C appears in 10 of those 17 molecules.  That seems to be a particularly difficult pattern for it.  =S=O appears in two others.

The most puzzling one is the Na+ I- ion pair.  That ought to be a very easy case for it to learn.  For comparison, K+ I- also has fairly large errors (from 25 to 91 kJ/mol), but Na+ Cl- has much lower errors (between 2.7 and 26 kJ/mol).  Possibly we just need to generate more training data for these simple ions, especially I-.

**Comment by peastman** [comment]
For what it's worth, the dataset contains 7215 occurrences of Cl-, but only 3936 of I-.  That easily might (or might not) account for the difference in accuracy.

**Comment by peastman** [comment]
Looking at the plots of energy vs. distance for different pairs of monatomic ions, most of them look something like this:

![image](https://user-images.githubusercontent.com/4379786/172033902-d3d29b72-f50e-4fb9-845f-01fb6036c59f.png)

It does a good job at short distances, but becomes significantly less accurate at longer distances.   I don't think this reflects a limitation of the model.  It uses a cutoff of 10 A, so it should have no trouble at only a 3 A separation.  I suspect it's just a lack of training data.  Some are a little better than others, but the basic pattern is the same for nearly all of them.

The two exceptions are K+/I- and Na+/I-, which it just completely failed to learn.

![image](https://user-images.githubusercontent.com/4379786/172034032-30a38a11-691c-44a9-afbb-f8e9edcba86d.png)

This too probably just reflects insufficient training data.  I think we'll want to add more data for these simple pairs, especially at larger separations.  They'll be super cheap to calculate since they're only two atoms each and there won't be very many of them.

**Comment by peastman** [comment]
I generated some more training data for ion pairs and trained a new model.  The result is enormously improved.

![image](https://user-images.githubusercontent.com/4379786/172669999-da302274-1da3-4584-af98-ceafb5d99fd5.png)

I'll open a PR adding a new dataset with lots of ion pairs.

**Comment by peastman** [comment]
I'd like to investigate other ways of handling charges.  Any suggestions for methods to try?  I know @jchodera is fond of the method described in https://www.nature.com/articles/s41467-020-20427-2.  It consists of these steps.

1. The neural network predicts an electronegativity for each atom.
2. Solve a system of equations to get a partial charge for each atom.
3. Include an explicit Coulomb interaction between the partial charges.

A problem with this method is that step 2 scales very badly with the number of atoms.  Instead of directly solving the system of equations, perhaps we could use an iterative solver and hardcode a small, fixed number of iterations?

What else would be good to try?

**Comment by giadefa** [comment]
What about predicting charge density in space as an option ?

**Comment by peastman** [comment]
How would that work?  Is there a paper describing it?

**Comment by giadefa** [comment]
No paper that I know. The idea would be to leave the short range to the NNP
and go to Ewald summation for long range. Would that be doable?

**Comment by peastman** [comment]
In principle, though only for periodic systems of course.  In a sense that's what the other method does.  If you predict partial charges for each atom, you can use PME for the long range part.  The main NNP is then learning just a correction to it.  It sounds like you're thinking of something a bit more complicated though, where you directly predict a charge distribution rather than just atomic charges?

**Comment by giadefa** [comment]
Yes, exactly. Partial charges are a trick.

**Comment by jchodera** [comment]
Model-building scripts and trained models are now made available in https://github.com/openmm/spice-models
